src/sign/generic.rs

```diff
@@ -0,0 +1,512 @@
+///
+/// This type cannot be used for computing signatures, as it does not implement
+/// any cryptographic primitives.  Instead, it is a generic representation that
+/// can be imported/exported or converted into a [`Sign`] (if the underlying
+/// cryptographic implementation supports it).
+///
+/// [`Sign`]: super::Sign
+pub enum SecretKey<B: AsRef<[u8]> + AsMut<[u8]>> {
+    /// An RSA/SHA256 keypair.
+    RsaSha256(RsaSecretKey<B>),
+
+    /// An ECDSA P-256/SHA-256 keypair.
+    ///
+    /// The private key is a single 32-byte big-endian integer.
+    EcdsaP256Sha256([u8; 32]),
+
+    /// An ECDSA P-384/SHA-384 keypair.
+    ///
+    /// The private key is a single 48-byte big-endian integer.
+    EcdsaP384Sha384([u8; 48]),
+
+    /// An Ed25519 keypair.
+    ///
+    /// The private key is a single 32-byte string.
+    Ed25519([u8; 32]),
+
+    /// An Ed448 keypair.
+    ///
+    /// The private key is a single 57-byte string.
+    Ed448([u8; 57]),
+}
+
+impl<B: AsRef<[u8]> + AsMut<[u8]>> SecretKey<B> {
+    /// The algorithm used by this key.
+    pub fn algorithm(&self) -> SecAlg {
+        match self {
+            Self::RsaSha256(_) => SecAlg::RSASHA256,
+            Self::EcdsaP256Sha256(_) => SecAlg::ECDSAP256SHA256,
+            Self::EcdsaP384Sha384(_) => SecAlg::ECDSAP384SHA384,
+            Self::Ed25519(_) => SecAlg::ED25519,
+            Self::Ed448(_) => SecAlg::ED448,
+        }
+    }
+
+    /// Serialize this key in the conventional DNS format.
+    ///
+    /// - For RSA, see RFC 5702, section 6.
+    /// - For ECDSA, see RFC 6605, section 6.
+    /// - For EdDSA, see RFC 8080, section 6.
+    pub fn into_dns(&self, w: &mut impl fmt::Write) -> fmt::Result {
+        writeln!(w, "Private-key-format: v1.2")?;
+        match self {
+            Self::RsaSha256(k) => {
+                writeln!(w, "Algorithm: 8 (RSASHA256)")?;
+                k.into_dns(w)
+            }
+
+            Self::EcdsaP256Sha256(s) => {
+                writeln!(w, "Algorithm: 13 (ECDSAP256SHA256)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::EcdsaP384Sha384(s) => {
+                writeln!(w, "Algorithm: 14 (ECDSAP384SHA384)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::Ed25519(s) => {
+                writeln!(w, "Algorithm: 15 (ED25519)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::Ed448(s) => {
+                writeln!(w, "Algorithm: 16 (ED448)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+        }
+    }
+
+    /// Parse a key from the conventional DNS format.
+    ///
+    /// - For RSA, see RFC 5702, section 6.
+    /// - For ECDSA, see RFC 6605, section 6.
+    /// - For EdDSA, see RFC 8080, section 6.
+    pub fn from_dns(data: &str) -> Result<Self, DnsFormatError>
+    where
```

Comment:
I intentionally left out a lot of information here in the same spirit as `ring::error::Unspecified` -- it could potentially leak information about the private key.  In any case, providing specific error messages is useless from an end-user perspective: there's an infinitesimal chance they actually opened the private key file and edited it by hand.  At best, any error messages visible to the end user should link to a specification of the format (not that I can find any).